src/server/scripts/EasternKingdoms/Stratholme/boss_baroness_anastari.cpp

```diff
@@ -16,112 +16,144 @@
 
-        uint32 BansheeWail_Timer;
-        uint32 BansheeCurse_Timer;
-        uint32 Silence_Timer;
-        //uint32 Possess_Timer;
+    EventMap _events;
+    ObjectGuid _possessedTargetGuid;
 
-        void Reset() override
+    void Reset() override
+    {
+        if (!me->IsVisible())
         {
-            Initialize();
+            if (Player* possessedTarget = ObjectAccessor::FindConnectedPlayer(_possessedTargetGuid))
+            {
+                if (possessedTarget->IsAlive())
+                {
+                    possessedTarget->RemoveAurasDueToSpell(SPELL_POSSESSED);
+                    possessedTarget->RemoveAurasDueToSpell(SPELL_POSSESS);
+                    me->SetReactState(REACT_AGGRESSIVE);
+                    me->SetVisible(true);
+                }
+            }
         }
+        _events.Reset();
+    }
 
-        void JustEngagedWith(Unit* /*who*/) override
-        {
-        }
+    void JustEngagedWith(Unit* /*who*/) override
+    {
+        _events.ScheduleEvent(EVENT_SPELL_BANSHEEWAIL, 1s);
+        _events.ScheduleEvent(EVENT_SPELL_BANSHEECURSE, 11s);
+        _events.ScheduleEvent(EVENT_SPELL_SILENCE, 13s);
+        _events.ScheduleEvent(EVENT_SPELL_POSSESS, 20s, 30s);
+    }
 
-        void JustDied(Unit* /*killer*/) override
-        {
-            instance->SetData(TYPE_BARONESS, IN_PROGRESS);
-        }
+    void JustDied(Unit* /*killer*/) override
+    {
+        instance->SetData(TYPE_BARONESS, IN_PROGRESS);
+    }
 
-        void UpdateAI(uint32 diff) override
-        {
-            if (!UpdateVictim())
-                return;
+    void UpdateAI(uint32 diff) override
+    {
+        if (!UpdateVictim())
+            return;
 
-            //BansheeWail
-            if (BansheeWail_Timer <= diff)
-            {
-                if (rand32() % 100 < 95)
-                    DoCastVictim(SPELL_BANSHEEWAIL);
-                //4 seconds until we should cast this again
-                BansheeWail_Timer = 4000;
-            } else BansheeWail_Timer -= diff;
+        _events.Update(diff);
 
-            //BansheeCurse
-            if (BansheeCurse_Timer <= diff)
-            {
-                if (rand32() % 100 < 75)
-                    DoCastVictim(SPELL_BANSHEECURSE);
-                //18 seconds until we should cast this again
-                BansheeCurse_Timer = 18000;
-            } else BansheeCurse_Timer -= diff;
+        if (me->HasUnitState(UNIT_STATE_CASTING))
+            return;
 
-            //Silence
-            if (Silence_Timer <= diff)
+        while (uint32 eventId = _events.ExecuteEvent())
+        {
+            switch (eventId)
             {
-                if (rand32() % 100 < 80)
-                    DoCastVictim(SPELL_SILENCE);
-                //13 seconds until we should cast this again
-                Silence_Timer = 13000;
-            } else Silence_Timer -= diff;
-
-            DoMeleeAttackIfReady();
+            case EVENT_SPELL_BANSHEEWAIL:
```

Comment:
`case` should be indented

```cpp
switch (expr)
{
    case LABEL:
        DoStuff();
```

Reply:
its in switch case? Or do I miss something?
```
while (uint32 eventId = _events.ExecuteEvent())
        {
            switch (eventId)
            {
            case EVENT_SPELL_BANSHEEWAIL:
                DoCastVictim(SPELL_BANSHEEWAIL);
                _events.ScheduleEvent(EVENT_SPELL_BANSHEEWAIL, 4s);
```

Reply:
you have
```cpp
switch(expr)
{
case LABEL:
    stuff();
```

should be
```cpp
switch (expr)
{
    case LABEL:
        stuff();
```